### PR TITLE
[mg] Implement Commands to Add/Remove an Environment, Login/Logout and Update Deploy/List/UnDeploy API

### PR DESCRIPTION
**Pull request by SuKSW**
## Purpose
Implement and update the commands mentioned in the title. Following will be the commands:

apictl mg add env
apictl mg remove env
apictl mg login
apictl mg logout
apictl mg deploy
apictl mg list
apictl mg undeploy

## Automation tests
 - Unit tests 
   > Code coverage information
 - Integration tests
   > Details about the test cases and coverage

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes/no
 - Ran FindSecurityBugs plugin and verified report? yes/no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes/no

## Related PRs
Closes https://github.com/wso2/product-apim-tooling/issues/621

### CODE REVIEW

File: import-export-cli/cmd/mg/logout.go [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+package mg
+
+import (
+	"github.com/spf13/cobra"
+	impl "github.com/wso2/product-apim-tooling/import-export-cli/impl/mg"
+	"github.com/wso2/product-apim-tooling/import-export-cli/utils"
+)
+
+const logoutCmdLiteral = "logout [environment]"
+const logoutCmdShortDesc = "Logout to from an Microgateway Adapter environment"
```
Comment (Chamindu36) [inline import-export-cli/cmd/mg/logout.go:28]: Shall we fix this small typo here?
```suggestion
const logoutCmdShortDesc = "Logout from an Microgateway Adapter environment"
```

File: import-export-cli/cmd/mg/removeEnv.go [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+		utils.Logln(utils.LogPrefixInfo + envCmdLiteral + " called")
+		executeRemoveEnvCmd(envToBeRemoved, utils.MainConfigFilePath)
+	},
+}
+
+func executeRemoveEnvCmd(env, mainConfigFilePath string) {
+	err := impl.RemoveEnv(env, mainConfigFilePath)
+	if err != nil {
+		utils.HandleErrorAndExit("Error removing environment", err)
```
Comment (Chamindu36) [inline import-export-cli/cmd/mg/removeEnv.go:58]: Shall we change this as follows?
```suggestion
		utils.HandleErrorAndExit("Error occurred when removing environment", err)
```

File: import-export-cli/impl/mg/removeEnv.go [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+)
+
+// removeEnv
+// to be used with 'remove env' command
+// @param envName : Name of the environment to be removed from the
+// @param mainConfigFilePath : Path to file where env endpoints are stored
+// @return error
+func RemoveEnv(envName, mainConfigFilePath string) error {
+	if envName == "" {
+		return errors.New("name of the environment cannot be blank")
```
Comment (Chamindu36) [inline import-export-cli/impl/mg/removeEnv.go:35]: Shall we change this as follows?
```suggestion
		return errors.New("Name of the environment cannot be blank")
```

File: import-export-cli/utils/envManagementUtils.go [hunk 4]
```diff
@@ -430,3 +430,51 @@ func APIMExistsInEnv(env, filePath string) bool {
...
+
+// Return EnvEndpoints for a given environment
+func GetEndpointsOfMgwAdapterEnv(env string, filePath string) (*MgwEndpoints, error) {
+	mainConfig := GetMainConfigFromFile(filePath)
+	for _env, mgwEndpoints := range mainConfig.MgwAdapterEnvs {
+		if _env == env {
+			return &mgwEndpoints, nil
+		}
+	}
+	return nil, errors.New("error getting endpoints of mgw adapter environment '" + env +
```
Comment (Chamindu36) [inline import-export-cli/utils/envManagementUtils.go:459]: Shall we change this as follows?
```suggestion
	return nil, errors.New("Error getting endpoints of mgw adapter environment '" + env +
```

File: import-export-cli/utils/envManagementUtils.go [hunk 5]
```diff
@@ -430,3 +430,51 @@ func APIMExistsInEnv(env, filePath string) bool {
...
+		"'. Try adding it using `apictl mg add env`")
+}
+
+// @param env : Environment to be removed from file
+// @param endpointsFilePath : Path to file where env endpoints are stored
+func RemoveMgwAdapterEnvFromMainConfigFile(env, endpointsFilePath string) error {
+	if env == "" {
+		return errors.New("environment cannot be blank")
```
Comment (Chamindu36) [inline import-export-cli/utils/envManagementUtils.go:467]: Shall we change this to **"Environment cannot be blank"**

File: import-export-cli/cmd/mg/login.go [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+	impl "github.com/wso2/product-apim-tooling/import-export-cli/impl/mg"
+	"github.com/wso2/product-apim-tooling/import-export-cli/utils"
+)
+
+var loginUsername string
+var loginPassword string
+var loginPasswordStdin bool
+
+const loginCmdLiteral = "login [environment]"
+const loginCmdShortDesc = "Login to an Microgateway Adapter environment"
```
Comment (wasuradananjith) [inline import-export-cli/cmd/mg/login.go:32]:
```suggestion
const loginCmdShortDesc = "Login to a Microgateway Adapter environment"
```
Shall we fix this typo?

File: import-export-cli/cmd/mg/login.go [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+const loginCmdLongDesc = `Login to an Microgateway Adapter environment using username and password`
```
Comment (wasuradananjith) [inline import-export-cli/cmd/mg/login.go:33]:
```suggestion
const loginCmdLongDesc = `Login to a Microgateway Adapter environment using username and password`
```
Here as well

File: import-export-cli/credentials/credentials.go [hunk 8]
```diff
@@ -47,6 +47,8 @@ type Credential struct {
 type Credentials struct {
 	// Environments specific credentials
 	Environments map[string]Environment `json:"environments"`
+	// Microgateway Adapter Environments specific credentials
+	MgwAdapterEnvs map[string]MgToken `json:"mgw_adapter_envs"`
```
Comment (uvindra) [inline import-export-cli/credentials/credentials.go:51]: Can we change `json:"mgw_adapter_envs"` to camel case to make it consistent with the way we are naming json keys elsewhere?

File: import-export-cli/utils/structs.go [hunk 9]
```diff
@@ -29,8 +29,9 @@ type EnvKeysAll struct {
 // For main_config.yaml
 // To be manually edited by the user
 type MainConfig struct {
-	Config       Config                  `yaml:"config"`
-	Environments map[string]EnvEndpoints `yaml:"environments"`
+	Config         Config                  `yaml:"config"`
+	Environments   map[string]EnvEndpoints `yaml:"environments"`
+	MgwAdapterEnvs map[string]MgwEndpoints `yaml:"mgw_adapter_envs"`
```
Comment (uvindra) [inline import-export-cli/utils/structs.go:34]: Can we change json:"mgw_adapter_envs" to camel case to make it consistent with the way we are naming json keys elsewhere?

File: import-export-cli/credentials/credentials.go (hunk 8 above)
Comment (SuKSW) [reply import-export-cli/credentials/credentials.go:51]:
```
config:
  http_request_timeout: 10000
  export_directory: /home/subhashinie/.wso2apictl/exported
  kubernetes_mode: false
  token_type: JWT
  vcs_deletion_enabled: false
  vcs_config_file_path: ""
  tls-renegotiation-mode: never
environments:
  dev:
    apim: someurl
    publisher: ""
    devportal: ""
    registration: ""
    admin: ""
    token: someurl/oauth2/token
    mi: ""
mgw_adapter_envs:
  dev:
    adapter: anotherurl
```
This is how the config looks like after adding envs.  Which is why I used this convention.

File: import-export-cli/cmd/mg/logout.go (hunk 1 above)
Comment (SuKSW) [reply import-export-cli/cmd/mg/logout.go:28]: Thanks, updating this.

File: import-export-cli/cmd/mg/removeEnv.go (hunk 2 above)
Comment (SuKSW) [reply import-export-cli/cmd/mg/removeEnv.go:58]: Thanks, updating this.

File: import-export-cli/impl/mg/removeEnv.go (hunk 3 above)
Comment (SuKSW) [reply import-export-cli/impl/mg/removeEnv.go:35]: Thanks, updating this.

File: import-export-cli/utils/envManagementUtils.go (hunk 4 above)
Comment (SuKSW) [reply import-export-cli/utils/envManagementUtils.go:459]: Thanks, updating this.

File: import-export-cli/utils/envManagementUtils.go (hunk 5 above)
Comment (SuKSW) [reply import-export-cli/utils/envManagementUtils.go:467]: Thanks, updating this.

File: import-export-cli/cmd/mg/login.go (hunk 6 above)
Comment (SuKSW) [reply import-export-cli/cmd/mg/login.go:32]: Thanks, updating this.

File: import-export-cli/cmd/mg/login.go (hunk 7 above)
Comment (SuKSW) [reply import-export-cli/cmd/mg/login.go:33]: Thanks, updating this.

File: import-export-cli/utils/structs.go (hunk 9 above)
Comment (SuKSW) [reply import-export-cli/utils/structs.go:34]: Same explanation as https://github.com/wso2/product-apim-tooling/pull/625#discussion_r580809405

**Review by Chamindu36 — APPROVED** [review]

**Review by uvindra — APPROVED** [review]

**Review by wasuradananjith — APPROVED** [review]

**Review by 1akshitha — APPROVED** [review]

**Pull request merged**